From bd0deb6a984c34a3bee1172ae0e3a986bba2b54f Mon Sep 17 00:00:00 2001
From: badrikesh prusty <badrikesh.prusty@siemens.com>
Date: Sun, 29 Jun 2025 15:42:50 -0400
Subject: [PATCH v4] isar-installer: Mount /tmp as tmpfs for read-only boot fix

Fix image installation when isar-image-installer rootfs is booted as read-only
by ensuring /tmp is mounted as tmpfs.

With the latest isar installer, Bmaptool now creates the /tmp/progress pipe to
track and display installation progress in a progress bar.

To mount /tmp as tmpfs, this change utilizes systemd's tmp.mount. As tmp.mount
is located in /usr/share/systemd/ for systemd versions up to Debian Bookworm,
and in /usr/lib/systemd/system/tmp.mount for versions 256.x and newer, the file
must be copied to /etc/systemd/system/.

Placing tmp.mount unit file in /etc/systemd/system/ ensures its configuration
for /tmp takes precedence, overriding any entry found in /etc/fstab.

Signed-off-by: badrikesh prusty <badrikesh.prusty@siemens.com>
---
 .../deploy-image/deploy-image_0.1.bb             |  6 ++++--
 .../systemd-tmpfs-tmp/files/postinst             | 16 ++++++++++++++++
 .../systemd-tmpfs-tmp/systemd-tmpfs-tmp_0.1.bb   | 13 +++++++++++++
 3 files changed, 33 insertions(+), 2 deletions(-)
 create mode 100644 meta-isar/recipes-installer/systemd-tmpfs-tmp/files/postinst
 create mode 100644 meta-isar/recipes-installer/systemd-tmpfs-tmp/systemd-tmpfs-tmp_0.1.bb

diff --git a/meta-isar/recipes-installer/deploy-image/deploy-image_0.1.bb b/meta-isar/recipes-installer/deploy-image/deploy-image_0.1.bb
index 54e521b5..b287a8d1 100644
--- a/meta-isar/recipes-installer/deploy-image/deploy-image_0.1.bb
+++ b/meta-isar/recipes-installer/deploy-image/deploy-image_0.1.bb
@@ -10,8 +10,10 @@ inherit dpkg-raw
 SRC_URI = "file://usr/bin/deploy-image-wic.sh \
            file://usr/lib/deploy-image-wic/handle-config.sh \
           "
-DEPENDS:bookworm += "bmap-tools"
-DEBIAN_DEPENDS = "bmap-tools, pv, dialog, util-linux, parted, fdisk, gdisk, pigz, xz-utils, pbzip2, zstd"
+
+DEPENDS:append:bookworm = " bmap-tools"
+DEPENDS:append = " systemd-tmpfs-tmp"
+DEBIAN_DEPENDS = "bmap-tools, pv, dialog, util-linux, parted, fdisk, gdisk, pigz, systemd-tmpfs-tmp, xz-utils, pbzip2, zstd"
 do_install[cleandirs] = "${D}/usr/bin/ \
                          ${D}/usr/lib/deploy-image-wic \
                         "
diff --git a/meta-isar/recipes-installer/systemd-tmpfs-tmp/files/postinst b/meta-isar/recipes-installer/systemd-tmpfs-tmp/files/postinst
new file mode 100644
index 00000000..9d75025d
--- /dev/null
+++ b/meta-isar/recipes-installer/systemd-tmpfs-tmp/files/postinst
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+set -e
+
+DEST="/etc/systemd/system/tmp.mount"
+
+if [ ! -e "${DEST}" ]; then
+    # Possible source paths for tmp.mount
+    for SRC in /usr/share/systemd/tmp.mount /usr/lib/systemd/system/tmp.mount; do
+        # If the source file exists, copy it to the destination and exit the loop
+        [ -e "${SRC}" ] && cp ${SRC} ${DEST} && break
+    done
+fi
+
+# Enable the mount unit using deb-systemd-helper; ignore errors if it fails
+deb-systemd-helper enable tmp.mount || true
diff --git a/meta-isar/recipes-installer/systemd-tmpfs-tmp/systemd-tmpfs-tmp_0.1.bb b/meta-isar/recipes-installer/systemd-tmpfs-tmp/systemd-tmpfs-tmp_0.1.bb
new file mode 100644
index 00000000..c0ab1bcf
--- /dev/null
+++ b/meta-isar/recipes-installer/systemd-tmpfs-tmp/systemd-tmpfs-tmp_0.1.bb
@@ -0,0 +1,13 @@
+# This software is a part of ISAR.
+# Copyright (C) Siemens AG, 2025
+#
+# SPDX-License-Identifier: MIT
+
+inherit dpkg-raw
+
+DESCRIPTION = "Configures /tmp as systemd-managed temporary filesystem (tmpfs), ensuring read-write access even if rootfs is read-only"
+MAINTAINER = "isar-users <isar-users@googlegroups.com>"
+
+SRC_URI = "file://postinst"
+
+DEBIAN_DEPENDS = "systemd"
-- 
2.47.2

